draft-ietf-masque-h3-datagram.md

```diff
@@ -117,26 +123,192 @@ following format (using the notation from the "Notational Conventions" section
 of {{QUIC}}):
 
 ~~~
-HTTP/3 DATAGRAM Frame {
-  Flow ID (i),
+HTTP/3 Datagram {
+  Quarter Stream ID (i),
+  Context ID (i),
   HTTP/3 Datagram Payload (..),
 }
 ~~~
 {: #h3-datagram-format title="HTTP/3 DATAGRAM Frame Format"}
 
-Flow ID:
+Quarter Stream ID:
+
+: A variable-length integer that contains the value of the client-initiated
+bidirectional stream that this datagram is associated with, divided by four.
+(The division by four stems from the fact that HTTP requests are sent on
+client-initiated bidirectional streams, and those have stream IDs that are
+divisible by four.)
+
+Context ID:
+
+: A variable-length integer indicating the context ID of the datagram (see
+{{datagram-contexts}}).
+
+HTTP/3 Datagram Payload:
+
+: The payload of the datagram, whose semantics are defined by individual
+applications. Note that this field can be empty.
+
+Intermediaries parse the Quarter Stream ID field in order to associate the QUIC
+DATAGRAM frame with a stream. If an intermediary receives a QUIC DATAGRAM frame
+whose payload is too short to allow parsing the Quarter Stream ID field, the
+intermediary MUST treat it as an HTTP/3 connection error of type
+H3_GENERAL_PROTOCOL_ERROR. Intermediaries MUST ignore any HTTP/3 Datagram
+fields after the Quarter Stream ID.
+
+Endpoints parse both the Quarter Stream ID field and the Context ID field in
+order to associate the QUIC DATAGRAM frame with a stream and context within
+that stream. If an endpoint receives a QUIC DATAGRAM frame whose payload is too
+short to allow parsing the Quarter Stream ID field, the endpoint MUST treat it
+as an HTTP/3 connection error of type H3_GENERAL_PROTOCOL_ERROR. If an endpoint
+receives a QUIC DATAGRAM frame whose payload is long enough to allow parsing
+the Quarter Stream ID field but too short to allow parsing the Context ID
+field, the endpoint MUST abruptly terminate the corresponding stream with a
+stream error of type H3_GENERAL_PROTOCOL_ERROR.
+
+
+# CAPSULE HTTP/3 Frame Definition {#capsule-frame}
```

Comment:
After reading this document, the main benefit I see to CAPSULE is that it makes it easier to deploy future end-to-end extension "frames" without having to modify intermediaries (because H3 drops unknown frames).  It may also make it easier to extend in client <--> server cases where the H3 stack itself is not easily modifiable (owned by a third-party, say).

Eventually it would probably warrant its own document and not be buried in DATAGRAM?

Reply:
I agree with your benefit analysis. I'd prefer to avoid splitting the document further though. Let's wait and see if anyone wants to use CAPSULE but cannot implement DATAGRAM.